synapse/app/generic_worker.py

```diff
@@ -863,9 +863,24 @@ async def update_token(self, token):
         a FEDERATION_ACK back to the master, and stores the token that we have processed
          in `federation_stream_position` so that we can restart where we left off.
         """
-        try:
-            self.federation_position = token
+        self.federation_position = token
+
+        # We save and send the ACK to master asynchronously, so we don't block
+        # processing on persistence. We don't need to do this operation for
+        # every single RDATA we receive, we just need to do it periodically.
+
+        if self._fed_position_linearizer.is_queued(None):
+            # There is already a task queued up to save and send the token, so
+            # no need to queue up another task.
+            return
 
+        run_as_background_process("_save_and_send_ack", self._save_and_send_ack)
+
+    async def _save_and_send_ack(self):
```

Comment:
Do you need to save `self.federation_position` locally for this function?

What happens if:

* This is running in the background and we await on `update_federation_out_pos`
* More updates come in so `self.federation_position` gets reved
* We then ack up to the "new" position, not what was saved...I think?

Reply:
Good point well made I like it.